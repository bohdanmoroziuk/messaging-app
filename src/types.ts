export interface Location {
  latitude: number;
  longitude: number;
}

export interface TextMessage {
  id: number;
  type: 'text';
  text: string;
}

export interface ImageMessage {
  id: number;
  type: 'image';
  uri: string;
}

export interface LocationMessage {
  id: number;
  type: 'location';
  location: Location;
}

export type Message =
 | TextMessage
 | ImageMessage
 | LocationMessage;
