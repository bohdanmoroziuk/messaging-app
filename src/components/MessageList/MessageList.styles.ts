import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  container: {
    flex: 1,
    overflow: 'visible',
  },
  messageRow: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    marginBottom: 4,
    marginRight: 10,
    marginLeft: 60,
  },
  messageBubble: {
    paddingVertical: 5,
    paddingHorizontal: 10,
    backgroundColor: 'rgb(16, 135, 255)',
    borderRadius: 20,
  },
  text: {
    fontSize: 18,
    color: 'white',
  },
  image: {
    width: 150,
    height: 150,
    borderRadius: 10,
  },
  map: {
    width: 250,
    height: 250,
    borderRadius: 10,
  },
});
