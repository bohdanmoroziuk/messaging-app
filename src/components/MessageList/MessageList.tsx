import { FunctionComponent } from 'react';
import { FlatList, Image, Text, View, TouchableOpacity } from 'react-native';
import MapView, { Marker as MapViewMarker } from 'react-native-maps';

import { Message } from 'src/types';

import styles from './MessageList.styles';

export interface MessageListProps {
  messages: Message[];
  onPressMessage: (message: Message) => void;
}

const keyExtractor = (message: Message) => message.id.toString();

export const MessageList: FunctionComponent<MessageListProps> = ({
  messages,
  onPressMessage,
}) => {
  const renderMessageBody = (message: Message) => {
    switch (message.type) {
      case 'text':
        return (
          <View style={styles.messageBubble}>
            <Text style={styles.text}>
              {message.text}
            </Text>
          </View>
        );
      case 'image':
        return (
          <Image
            style={styles.image}
            source={{ uri: message.uri }}
          />
        );
      case 'location':
        return (
          <MapView
            style={styles.map}
            initialRegion={{
              ...message.location,
              latitudeDelta: 0.08,
              longitudeDelta: 0.04,
            }}
          >
            <MapViewMarker coordinate={message.location} />
          </MapView>
        );
      default:
        return null;
    }
  };

  const renderMessageItem = ({ item }: { item: Message }) => {
    const handlePress = () => {
      onPressMessage(item);
    };

    return (
      <View
        key={item.id}
        style={styles.messageRow}
      >
        <TouchableOpacity onPress={handlePress}>
          {renderMessageBody(item)}
        </TouchableOpacity>
      </View>
    );
  };

  return (
    <FlatList
      style={styles.container}
      inverted
      data={messages}
      renderItem={renderMessageItem}
      keyExtractor={keyExtractor}
      keyboardShouldPersistTaps="handled"
    />
  );
};
