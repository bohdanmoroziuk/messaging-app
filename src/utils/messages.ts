import { Message, Location } from 'src/types';

let messageId = 0;

export const getNextId = () => {
  messageId += 1;
  return messageId;
};

export const createTextMessage = (text: string): Message => ({
  id: getNextId(),
  type: 'text',
  text,
});

export const createImageMessage = (uri: string): Message => ({
  id: getNextId(),
  type: 'image',
  uri,
});

export const createLocationMessage = (location: Location): Message => ({
  id: getNextId(),
  type: 'location',
  location,
});
