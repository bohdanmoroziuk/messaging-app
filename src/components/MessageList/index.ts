import { MessageList } from './MessageList';

export type { MessageListProps } from './MessageList';
export default MessageList;
