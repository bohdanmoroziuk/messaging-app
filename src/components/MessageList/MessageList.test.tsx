import { fireEvent, render } from '@testing-library/react-native';

import { MessageList } from './MessageList';
import { createTextMessage, createImageMessage, createLocationMessage } from '../../utils/messages';
import { Message } from '../../types';

describe('MessageList', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const mockMessages: Message[] = [
    createImageMessage('https://unsplash.it/300/300'),
    createTextMessage('Hello'),
    createTextMessage('World'),
    createLocationMessage({
      latitude: 37.78825,
      longitude: -122.4324,
    }),
  ];

  const mockOnPressMessage = jest.fn();

  it('renders successfully when the message list is empty', () => {
    const { toJSON } = render(
      <MessageList
        messages={[]}
        onPressMessage={mockOnPressMessage}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders successfully when the message list contains items', () => {
    const { toJSON } = render(
      <MessageList
        messages={mockMessages}
        onPressMessage={mockOnPressMessage}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('calls the onPressMessage handler when a list item is pressed', () => {
    const { getByText } = render(
      <MessageList
        messages={mockMessages}
        onPressMessage={mockOnPressMessage}
      />
    );

    const message = getByText('Hello');

    expect(message).toBeDefined();

    fireEvent.press(message);

    expect(mockOnPressMessage).toHaveBeenCalled();
  });
});
