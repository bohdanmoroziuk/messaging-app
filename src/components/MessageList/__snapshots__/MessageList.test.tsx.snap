// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList renders successfully when the message list contains items 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "id": 1,
        "type": "image",
        "uri": "https://unsplash.it/300/300",
      },
      Object {
        "id": 2,
        "text": "Hello",
        "type": "text",
      },
      Object {
        "id": 3,
        "text": "World",
        "type": "text",
      },
      Object {
        "id": 4,
        "location": Object {
          "latitude": 37.78825,
          "longitude": -122.4324,
        },
        "type": "location",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  invertStickyHeaders={true}
  inverted={true}
  keyExtractor={[Function]}
  keyboardShouldPersistTaps="handled"
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "transform": Array [
          Object {
            "scaleY": -1,
          },
        ],
      },
      Object {
        "flex": 1,
        "overflow": "visible",
      },
    ]
  }
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "marginBottom": 4,
            "marginLeft": 60,
            "marginRight": 10,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://unsplash.it/300/300",
              }
            }
            style={
              Object {
                "borderRadius": 10,
                "height": 150,
                "width": 150,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "marginBottom": 4,
            "marginLeft": 60,
            "marginRight": 10,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(16, 135, 255)",
                "borderRadius": 20,
                "paddingHorizontal": 10,
                "paddingVertical": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontSize": 18,
                }
              }
            >
              Hello
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "marginBottom": 4,
            "marginLeft": 60,
            "marginRight": 10,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(16, 135, 255)",
                "borderRadius": 20,
                "paddingHorizontal": 10,
                "paddingVertical": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontSize": 18,
                }
              }
            >
              World
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "marginBottom": 4,
            "marginLeft": 60,
            "marginRight": 10,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <AIRMap
            handlePanDrag={false}
            initialRegion={
              Object {
                "latitude": 37.78825,
                "latitudeDelta": 0.08,
                "longitude": -122.4324,
                "longitudeDelta": 0.04,
              }
            }
            onChange={[Function]}
            onLayout={[Function]}
            onMapReady={[Function]}
            onMarkerPress={[Function]}
            region={null}
            style={
              Object {
                "borderRadius": 10,
                "height": 250,
                "width": 250,
              }
            }
          >
            <AIRMapMarker
              coordinate={
                Object {
                  "latitude": 37.78825,
                  "longitude": -122.4324,
                }
              }
              onPress={[Function]}
              stopPropagation={false}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "position": "absolute",
                  },
                  undefined,
                ]
              }
            />
          </AIRMap>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`MessageList renders successfully when the message list is empty 1`] = `
<RCTScrollView
  data={Array []}
  getItem={[Function]}
  getItemCount={[Function]}
  invertStickyHeaders={true}
  inverted={true}
  keyExtractor={[Function]}
  keyboardShouldPersistTaps="handled"
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "transform": Array [
          Object {
            "scaleY": -1,
          },
        ],
      },
      Object {
        "flex": 1,
        "overflow": "visible",
      },
    ]
  }
  viewabilityConfigCallbackPairs={Array []}
>
  <View />
</RCTScrollView>
`;
